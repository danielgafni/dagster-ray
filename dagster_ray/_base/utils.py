from __future__ import annotations

import os

from dagster import AssetExecutionContext, InitResourceContext, OpExecutionContext

from dagster_ray._base.constants import DEFAULT_DEPLOYMENT_NAME


def get_dagster_tags(
    context: InitResourceContext | OpExecutionContext | AssetExecutionContext, extra_tags: dict[str, str] | None = None
) -> dict[str, str]:
    """
    Returns a dictionary with common Dagster tags.
    """
    assert context.dagster_run is not None

    labels: dict[str, str] = {
        "dagster.io/deployment": DEFAULT_DEPLOYMENT_NAME,  # TODO: this should come from extra_tags,
        **(extra_tags or {}),
    }

    if isinstance(context, InitResourceContext):
        labels.update(
            **context.dagster_run.dagster_execution_info,
        )

        for resource_key, resource_def in context.all_resource_defs.items():
            if resource_def is context.resource_def:
                # inject the resource key used for this KubeRay resource
                # this enables e.g reusing the same `RayCluster` across Dagster steps that require it without recreating the cluster
                labels["dagster.io/resource_key"] = resource_key
    else:
        labels.update(
            **context.run.dagster_execution_info,
        )

    if context.dagster_run.tags.get("user"):
        labels["dagster.io/user"] = context.dagster_run.tags["user"]

    if os.getenv("DAGSTER_CLOUD_GIT_BRANCH"):
        labels["dagster.io/git-branch"] = os.environ["DAGSTER_CLOUD_GIT_BRANCH"]

    if os.getenv("DAGSTER_CLOUD_GIT_SHA"):
        labels["dagster.io/git-sha"] = os.environ["DAGSTER_CLOUD_GIT_SHA"]

    return labels
