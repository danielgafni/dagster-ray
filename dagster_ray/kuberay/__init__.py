from dagster_ray.kuberay.configs import RayClusterConfig
from dagster_ray.kuberay.jobs import cleanup_kuberay_clusters, delete_kuberay_clusters
from dagster_ray.kuberay.ops import cleanup_kuberay_clusters_op, delete_kuberay_clusters_op
from dagster_ray.kuberay.resources import KubeRayAPI, KubeRayCluster
from dagster_ray.kuberay.schedules import cleanup_kuberay_clusters_daily

__all__ = [
    "KubeRayCluster",
    "RayClusterConfig",
    "KubeRayAPI",
    "cleanup_kuberay_clusters",
    "delete_kuberay_clusters",
    "cleanup_kuberay_clusters_op",
    "delete_kuberay_clusters_op",
    "cleanup_kuberay_clusters_daily",
]
