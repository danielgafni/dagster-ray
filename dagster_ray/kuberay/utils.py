import re


def normalize_k8s_label_values(labels: dict[str, str]) -> dict[str, str]:
    # prevent errors like:
    # Invalid value: \"daniel@anam.ai\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')

    cleanup_regex = re.compile(r"[^a-zA-Z0-9-_.]+")

    for key, value in labels.items():
        # daniel~!@my.domain -> daniel-my-domain
        labels[key] = cleanup_regex.sub("", value.replace("@", "-").replace(".", "-"))

    return labels
