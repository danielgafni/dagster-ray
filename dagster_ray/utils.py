import os
from typing import Dict, List, Optional


def resolve_env_vars_list(env_vars: Optional[List[str]]) -> Dict[str, str]:
    res = {}

    if env_vars is not None:
        for env_var in env_vars:
            if "=" in env_var:
                var, value = env_var.split("=", 1)
                res[var] = value
            else:
                if value := os.getenv(env_var):
                    res[env_var] = value

    return res
