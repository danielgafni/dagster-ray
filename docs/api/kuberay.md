# KubeRay API Reference

KubeRay integration components for running Ray on Kubernetes.

::: dagster_ray.kuberay
    options:
      show_docstring_classes: true

## Client Mode Resources

These resources initialize Ray client connection with a remote cluster.

::: dagster_ray.kuberay.KubeRayInteractiveJob
    options:
      members:
        - "__init__"
        - "lifecycle"
        - "ray_job"
        - "client"
        - "log_cluster_conditions"


::: dagster_ray.kuberay.KubeRayCluster
    options:
      members:
        - "__init__"
        - "lifecycle"
        - "ray_cluster"
        - "client"
        - "log_cluster_conditions"

## Job Submission Resources

These resources submit Ray jobs to a remote cluster.

::: dagster_ray.kuberay.PipesKubeRayJobClient
    options:
      members:
        - "__init__"
        - "run"

## Configuration and Types

::: dagster_ray.kuberay.configs.RayJobConfig
    options:
      members:
        - "metadata"
        - "spec"
        - "to_k8s"

::: dagster_ray.kuberay.configs.RayJobSpec
    options:
      members:
        - "entrypoint"
        - "runtime_env"
        - "ray_cluster_spec"
        - "entrypoint_num_cpus"
        - "entrypoint_num_gpus"
        - "entrypoint_memory"
        - "entrypoint_resources"
        - "to_k8s"

---

::: dagster_ray.kuberay.resources.rayjob.InteractiveRayJobConfig
    options:
      members:
        - "metadata"
        - "spec"
        - "to_k8s"

::: dagster_ray.kuberay.resources.rayjob.InteractiveRayJobSpec
    options:
      inherited_members: true
      members: true


---

::: dagster_ray.kuberay.configs.RayClusterConfig
    options:
      members:
        - "metadata"
        - "spec"
        - "to_k8s"

::: dagster_ray.kuberay.configs.RayClusterSpec
    options:
      inherited_members: true
      members: true

--

::: dagster_ray.kuberay.resources.base.BaseKubeRayResourceConfig
    options:
      members:
        - "image"
        - "deployment_name"
        - "poll_interval"

## Resources

::: dagster_ray.kuberay.KubeRayJobClientResource
    options:
      members:
        - "__init__"

::: dagster_ray.kuberay.KubeRayClusterClientResource
    options:
      members:
        - "__init__"

## Kubernetes API Clients

::: dagster_ray.kuberay.client.RayClusterClient
    options:
      members:
        - "__init__"
        - "create"
        - "delete"
        - "get"
        - "list"
        - "update"

::: dagster_ray.kuberay.client.RayJobClient
    options:
      members:
        - "__init__"
        - "create"
        - "delete"
        - "get"
        - "list"
        - "update"
