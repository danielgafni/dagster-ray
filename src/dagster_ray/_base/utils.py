from __future__ import annotations

import dagster as dg

from dagster_ray._base.constants import DEFAULT_DEPLOYMENT_NAME


def get_dagster_tags(
    context: dg.InitResourceContext | dg.OpExecutionContext | dg.AssetExecutionContext,
    extra_tags: dict[str, str] | None = None,
) -> dict[str, str]:
    """
    Returns a dictionary with common Dagster tags.
    """
    assert context.dagster_run is not None

    labels: dict[str, str] = {
        "dagster/deployment": DEFAULT_DEPLOYMENT_NAME,  # TODO: this should come from extra_tags,
        **(extra_tags or {}),
    }

    if isinstance(context, dg.InitResourceContext):
        labels.update(
            **context.dagster_run.dagster_execution_info,
        )

        for resource_key, resource_def in context.all_resource_defs.items():
            if resource_def is context.resource_def:
                # inject the resource key used for this KubeRay resource
                # this enables e.g reusing the same `RayCluster` across Dagster steps that require it without recreating the cluster
                labels["dagster/resource-key"] = resource_key
    else:
        labels.update(
            **context.run.dagster_execution_info,
        )

    return labels
