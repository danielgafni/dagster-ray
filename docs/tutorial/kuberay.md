# Dagster + KubeRay

This tutorial shows how to use dagster-ray with KubeRay to automatically manage Ray clusters on Kubernetes. KubeRay integration allows you to create and manage Ray clusters directly from your Dagster pipelines without manual cluster management.

## Prerequisites

Before getting started, you'll need:

- A Kubernetes cluster with KubeRay Operator installed
- A `kubectl` configured to access your cluster or a kubeconfig file (resources can be configured to use it)
- dagster-ray installed with KubeRay support:
  ```bash
  pip install 'dagster-ray[kuberay]'
  ```

## KubeRayInteractiveJob

[`KubeRayInteractiveJob`](../api/kuberay.md#dagster_ray.kuberay.KubeRayInteractiveJob) is the recommended way to run Ray workloads with automatic cluster management. It creates a `RayJob`, connects to it in client mode and sets the `jobId` field. Cleanup is handled by the KubeRay controller or by the resource lifecycle logic.


!!! warning

    KubeRay Operator 1.3.0 is required for this feature.

### Basic Example

Here's a simple example that creates a Ray cluster and runs a distributed computation:

```python
import dagster as dg
from dagster_ray.kuberay import KubeRayInteractiveJob, RayResource
import ray


@ray.remote
def sum_of_squares_in_slice(start: int, end: int) -> int:
    return sum(i**2 for i in range(start, end))


@dg.asset
def sum_of_squares(ray_cluster: RayResource) -> int:
    # Split work across workers
    num_workers = 4
    chunk_size = 1000 // num_workers

    futures = [
        sum_of_squares_in_slice.remote(i * chunk_size + 1, (i + 1) * chunk_size + 1)
        for i in range(num_workers)
    ]

    # Sum results from all workers
    return sum(ray.get(futures))


definitions = dg.Definitions(
    assets=[compute_sum_of_squares], resources={"ray_cluster": KubeRayInteractiveJob()}
)
```

!!! note
    `RayResource` is the common interface for all `dagster-ray` Ray resource which can be used as backend-agnostic type annotation

By default, the image will be inherited from the `dagster/image` Run tag. Alternatively, you can specify it using the `image` parameter.

`RayJob`'s `.metadata.name` will be generated automatically if not provided.

### Advanced Configuration

You can customize the Ray cluster configuration:

```python
from dagster_ray.kuberay import KubeRayInteractiveJob
from dagster_ray.kuberay.configs import RayJobConfig, RayJobSpec, RayClusterSpec

ray_cluster = KubeRayInteractiveJob(
    ray_job=RayJobConfig(
        metadata={
            "namespace": "my-custom-namespace",
            "labels": {"team": "my-team"},
            "annotations": {"example": "annotation"},
        },
        spec=RayJobSpec(
            ttl_seconds_after_finished=3600,
            deletion_strategy={
                "onSuccess": {"policy": "DeleteSelf"},
                "onFailure": {"policy": "DeleteSelf"},
            },
            ray_cluster_spec=RayClusterSpec(
                worker_group_specs=[
                    {
                        "groupName": "workers",
                        "replicas": 0,
                        "minReplicas": 0,
                        "maxReplicas": 10,
                        "rayStartParams": {},
                        "template": {
                            "metadata": {"labels": {}, "annotations": {}},
                            "spec": {
                                "imagePullSecrets": [],
                                "containers": [
                                    {
                                        "volumeMounts": [],
                                        "name": "worker",
                                        "imagePullPolicy": "Always",
                                    }
                                ],
                                "volumes": [],
                                "affinity": {},
                                "tolerations": [],
                                "nodeSelector": {},
                            },
                        },
                    }
                ]
            ),
        ),
    ),
    lifecycle=Lifecycle(cleanup="always"),
    timeout=600.0,
)
```

### KubeRayCluster Alternative

While [`KubeRayInteractiveJob`](../api/kuberay.md#dagster_ray.kuberay.KubeRayInteractiveJob) is recommended for most use cases, you can also use [`KubeRayCluster`](../api/kuberay.md#dagster_ray.kuberay.KubeRayCluster) for more persistent clusters:

```python
from dagster_ray.kuberay import KubeRayCluster
from dagster_ray.kuberay.configs import RayClusterConfig

# KubeRayCluster creates a persistent RayCluster CR
ray_cluster = KubeRayCluster(
    ray_cluster=RayClusterConfig(
        # Cluster configuration
        metadata={"name": "my-persistent-cluster"},
    )
)
```

!!! note
    `KubeRayCluster` is generally a weaker alternative to `KubeRayInteractiveJob` because:

    - It creates persistent clusters that may not get cleaned up properly, for example if something happens to the Dagster pod
    - It lacks `RayJob`'s features such as timeouts and existing cluster selection

## PipesKubeRayJobClient

[`PipesKubeRayJobClient`](../api/kuberay.md#dagster_ray.kuberay.PipesKubeRayJobClient) allows you to submit external Python scripts as Ray jobs with automatic cluster management. This is ideal when you want to decouple your Ray workload from your Dagster orchestration code or Python environment.

### Basic Pipes Example

First, create a Ray script that will run on the cluster:

```python title="ray_workload.py"
# ml_training.py - External Ray script
import ray
from dagster_pipes import open_dagster_pipes


@ray.remote
def train_ml_model(partition_id: int):
    """Dummy ML training function."""
    import time

    time.sleep(1)  # Simulate work
    return {"partition_id": partition_id, "accuracy": 0.95}


def main():
    with open_dagster_pipes() as context:
        context.log.info("Starting distributed ML training")

        # Get configuration from Dagster
        num_partitions = context.get_extra("num_partitions", 4)

        # Submit training jobs
        futures = [train_ml_model.remote(i) for i in range(num_partitions)]
        results = ray.get(futures)

        context.log.info(f"Training complete on {len(results)} partitions")

        accuracy = sum(result["accuracy"] for result in results) / len(results)

        # Report results
        context.report_asset_materialization(
            metadata={"num_partitions": len(results), "accuracy": accuracy},
            data_version="alpha",
        )


if __name__ == "__main__":
    main()
```

Now create a Dagster asset that uses `PipesKubeRayJobClient`:

```python
import dagster as dg
from dagster_ray.kuberay import PipesKubeRayJobClient


class MLTrainingConfig(dg.Config):
    num_partitions: int = 4


@dg.asset
def distributed_computation(
    context: dg.AssetExecutionContext,
    ray_pipes_client: PipesKubeRayJobClient,
) -> None:
    """Run distributed computation using Pipes + KubeRay."""

    # Submit the external Ray script
    return ray_pipes_client.run(
        context=context,
        ray_job={
            "entrypoint": "python ray_workload.py",
            "runtime_env": {
                "pip": ["dagster-pipes", "torch"],  # (1)!
            },
            "entrypoint_num_cpus": 1.0,
            "entrypoint_memory": 2 * 1024 * 1024 * 1024,  # 2GB
        },
        extras={
            "num_partitions": config.num_partitions,
        },
    )


definitions = dg.Definitions(
    assets=[distributed_computation],
    resources={"ray_pipes_client": PipesKubeRayJobClient()},
)
```

1. :bulb: `dagster-pipes` have to be installed in the remote environment!

When materializing the asset, the `PipesKubeRayJobClient` will submit the script as a `RayJob` custom resource, monitor its status, and stream back logs and Dagster metadata.
