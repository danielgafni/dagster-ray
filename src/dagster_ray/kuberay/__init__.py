from dagster_ray.kuberay.configs import RayClusterConfig, RayClusterSpec, RayJobConfig, RayJobSpec
from dagster_ray.kuberay.jobs import cleanup_kuberay_clusters, delete_kuberay_clusters
from dagster_ray.kuberay.ops import cleanup_kuberay_clusters_op, delete_kuberay_clusters_op
from dagster_ray.kuberay.pipes import PipesKubeRayJobClient
from dagster_ray.kuberay.resources import (
    KubeRayCluster,
    KubeRayClusterClientResource,
    KubeRayInteractiveJob,
    KubeRayJobClientResource,
)
from dagster_ray.kuberay.schedules import cleanup_kuberay_clusters_daily

__all__ = [
    "KubeRayCluster",
    "RayClusterConfig",
    "KubeRayClusterClientResource",
    "PipesKubeRayJobClient",
    "cleanup_kuberay_clusters",
    "delete_kuberay_clusters",
    "cleanup_kuberay_clusters_op",
    "delete_kuberay_clusters_op",
    "cleanup_kuberay_clusters_daily",
    "RayClusterSpec",
    "RayJobConfig",
    "RayJobSpec",
    "KubeRayInteractiveJob",
    "KubeRayJobClientResource",
]
