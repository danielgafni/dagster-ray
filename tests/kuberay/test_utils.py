from dagster_ray.kuberay.utils import normalize_k8s_label_values


def test_normalize_k8s_label_values():
    assert normalize_k8s_label_values(
        {
            "foo": "bar",
            "my/label": "my/value",
            "user": "daniel@my.org",
            "user-dirty": "daniel!`~@my.org",
            "alphanumeric": "abc123",
            "long": 64 * "a",
        }
    ) == {
        "foo": "bar",
        "my/label": "myvalue",
        "user": "daniel-my-org",
        "user-dirty": "daniel-my-org",
        "alphanumeric": "abc123",
        "long": 63 * "a",
    }
