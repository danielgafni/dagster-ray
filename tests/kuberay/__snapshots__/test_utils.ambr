# serializer version: 1
# name: test_normalize_k8s_label_values
  dict({
    'alphanumeric': 'abc123',
    'badstart': 'foo',
    'badstart_after_initial_replace': 'at-foo',
    'foo': 'bar',
    'long': 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
    'my/label': 'my-value',
    'user': 'daniel-at-my.org',
    'user-dirty': 'daniel-at-my.org',
  })
# ---
